File: java/src/main/java/freezingoctosansa/RingBuffer.java
package freezingoctosansa;

public class RingBuffer {

  private final Integer _size;
  private final Integer[] _buffer;
  private int _readPos;
  private int _writePos;

  public RingBuffer(int size) {
    _size = size;
    _buffer = new Integer[size];
    _readPos = -1;
    _writePos = 0;
  }

  public Integer read() {
    if (_readPos == -1) return null;
    Integer result = _buffer[_readPos];
    _readPos = (_readPos + 1) % _size;
    if (_readPos == _writePos) {
      _readPos = -1;
    }
    return result;
  }

  public void write(Integer item) {
    _buffer[_writePos] = item;
    if (_readPos == -1) {
      _readPos = _writePos;
    } else if (_readPos == _writePos) {
      _readPos = (_readPos + 1) % _size;
    }
    _writePos = (_writePos + 1) % _size;
  }
}

File: python/ringbuffer.py


class RingBuffer:

    def __init__(self, size):
        self.size = size
        self.buffer = [None]*size
        self.read_pos = -1
        self.write_pos = 0

    def read(self):
        if self.read_pos == -1:
            return None
        r = self.buffer[self.read_pos]
        self.read_pos = (self.read_pos + 1) % self.size
        if self.read_pos == self.write_pos:
            self.read_pos = -1
        return r

    def write(self, x):
        self.buffer[self.write_pos] = x
        if self.read_pos == -1:
            # first write on empty buffer
            self.read_pos = self.write_pos
        elif self.read_pos == self.write_pos:
            # write eats into read buffer
            self.read_pos = (self.read_pos + 1) % self.size
        # advance the write pointer
        self.write_pos = (self.write_pos + 1) % self.size


if __name__ == "__main__":
    ringBuffer = RingBuffer(2)
    assert ringBuffer.read() is None
    ringBuffer.write(2)
    assert ringBuffer.read() == 2
    assert ringBuffer.read() is None
    ringBuffer.write(3)
    assert ringBuffer.read() == 3
    assert ringBuffer.read() is None

    ringBuffer.write(4)
    ringBuffer.write(5)
    assert ringBuffer.read() == 4
    assert ringBuffer.read() == 5
    assert ringBuffer.read() is None

    ringBuffer.write(6)
    ringBuffer.write(7)
    ringBuffer.write(8)
    assert ringBuffer.read() == 7
    assert ringBuffer.read() == 8
    assert ringBuffer.read() is None

    ringBuffer.write(9)
    ringBuffer.write(10)
    ringBuffer.write(11)
    ringBuffer.write(12)
    assert ringBuffer.read() == 11
    assert ringBuffer.read() == 12
    ringBuffer.read() is None
    ringBuffer.read() is None
